### PR TITLE
Refactor badge color functions

### PR DESCRIPTION
**Pull request by paulmelnikow**
- Replace the idea of color schemes with the idea of named colors (since none of our colorschemes have used `colorA`)
- Pass through the normalized color to `_shields_test` to harmonize with BaseService and simplify testing
    - Update service tests
- Move responsibility for color generation into the npm package
- Remove several color helper functions and their tests
- Update gh-badge public API to accept `color` and `labelColor`

This is a precursor to refactoring some of the logo code for #2473.

### CODE REVIEW

**Comment by chris48s** [comment]
I've had a go at reviewing this. I'll leave a few comments but it probably needs another pass.

One of the issue I'm hitting trying to test is that editing stuff in the modal popups in the front end is throwing:

```
TypeError: Cannot read property 'toLowerCase' of null
at isHexColor (shields/gh-badges/lib/color.js:22:31)
```

which is making it difficult to test overrides, logos etc

File: gh-badges/lib/badge-cli.js [hunk 1]
```diff
@@ -4,15 +4,15 @@
 
 const makeBadge = require('./make-badge')
 const svg2img = require('./svg-to-img')
-const colorscheme = require('./colorscheme.json')
+const { namedColors } = require('./color')
 
 if (process.argv.length < 4) {
   console.log('Usage: badge subject status [:colorscheme] [.output] [@style]')
```
Comment (chris48s) [inline gh-badges/lib/badge-cli.js]: Does `[:colorscheme]` want updating here?

File: services/base.js [hunk 2]
```diff
@@ -339,6 +334,14 @@ class BaseService {
       label: defaultLabel,
     } = this.defaultBadgeData
 
+    let color
+    if (isError) {
+      // Disregard the override color.
+      color = coalesce(serviceColor, defaultColor, 'lightgrey')
+    } else {
+      color = coalesce(overrideColor, serviceColor, defaultColor, 'lightgrey')
+    }
```
Comment (chris48s) [inline services/base.js]: This section will want updating to account for #2740

File: services/dub/dub.tester.js [hunk 3]
```diff
@@ -3,19 +3,18 @@
...
-  colorScheme.red,
-  colorScheme.yellow,
-  colorScheme.yellowgreen,
-  colorScheme.green,
-  colorScheme.brightgreen
+  'red',
+  'yellow',
+  'yellowgreen',
+  'green',
+  'brightgreen'
```
Comment (chris48s) [inline services/dub/dub.tester.js]: This will need updating to account for #2751 - I moved/renamed some stuff

**Comment by chris48s** [comment]
It seems to be a bit inconsistent exactly which keypresses throw it, but here's some screen capture from me filling some text into form fields - You will need to fullscreen the GIF to see the details:

gifrecord_2019-01-13_184256

Does that help? (This is with `HEAD` of my branch at b60b8fb )

**Comment by paulmelnikow** [comment]
Gotcha. Yes! Thanks a bunch. Looks like a backend issue. I'll see if I can't track that down.

File: gh-badges/lib/color.js [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+  orange: '#fe7d37',
+  red: '#e05d44',
+  blue: '#007ec6',
+  grey: '#555',
+  gray: '#555',
+  lightgrey: '#9f9f9f',
+  lightgray: '#9f9f9f',
+}
+
+const hexColorRegex = /^([\da-f]{3}){1,2}$/i
```
Comment (calebcartwright) [inline gh-badges/lib/color.js]: do we need to add an optional `#` to this?

Comment (calebcartwright) [reply gh-badges/lib/color.js]: after looking through the tests I'm guessing the answer is no since `normalizeColor` handles this?

Comment (paulmelnikow) [reply gh-badges/lib/color.js]: I thought the same thing. However `isCSSColor('#ccc')` is true, so we don't need to accept it here.

Comment (calebcartwright) [reply gh-badges/lib/color.js]: Ahhh, thanks for that. I was scratching my head on that one

**Comment by calebcartwright** [comment]
Should we also run these changes against all service tests to validate there's no unexpected color change behavior?

**Comment by paulmelnikow** [comment]
> Should we also run these changes against all service tests to validate there's no unexpected color change behavior?

Yea, I did the full service test run in a number of the earlier builds, and cleaned up all the failures that seemed to be related. It's possible one or two have crept in since then, though at this point I think we can catch them more easily just after merging.

**Comment by paulmelnikow** [comment]
> Gotcha. Yes! Thanks a bunch. Looks like a backend issue. I'll see if I can't track that down.

@chris48s This issue is fixed, by the way.

**Comment by calebcartwright** [comment]
> Yea, I did the full service test run in a number of the earlier builds 

Ah sorry, I missed that

**Comment by calebcartwright** [comment]
Tangential question, but have we ever evaluated the feasibility of having Coveralls chime in on PRs with info about coverage changes? I imagine that could get tricky given the different test runs we have (and I know those coverage bots can get a little spammy on some PRs), but I think that could be valuable in some cases, especially for the core code

**Comment by paulmelnikow** [comment]
There was a little discussion in #1584 and I wrote a bit the other day in https://github.com/badges/shields/pull/2756#discussion_r247362497.

I think ideally we have two suites:

1. Core
    - Only runs the core tests
    - Is expected to climb toward 100%
    - Is reported in PRs
2. Services
     - Runs the service tests specified in PRs
     - Gets reported in PRs
     - Ideally, only includes code belonging to the services being run. In some cases it may be difficult to detect which service code should be included. If we run `[pypidownloads]` then which files in `pypi` should be covered? I suppose if we run a subset of a service's tests we could err on the side of under-reporting.

1 seems easier to build. We could start there. Ideally if we can find a way to hack this with Coveralls, I think we should do that… it'd involve somehow keeping the two coverage runs separate. It might be possible if we switch `daily-tests` to think it's reporting on a branch, though I'm not totally sure if that would work.

**Comment by calebcartwright** [comment]
I agree with targetting/emphasizing the 1. Core option 

I suppose we could leverage multiple .nycrc files, and perhaps another option would be to use Coveralls for one run and maybe a different service like Codecov for another? I'd venture to say others have encountered a similar scenario before (for example monorepos) and wanted coverage on different parts so I'd be curious to see what others have done.

**Comment by calebcartwright** [comment]
Also as I read through #1584 I realize I am just repeating things you've all already thought of 😆

**Comment by calebcartwright** [comment]
While searching for Coveralls docs I found this from codecov: https://docs.codecov.io/docs/flags. I'll see what I can find on Coveralls and will open another issue if need be to avoid taking this PR any more off topic!

**Comment by paulmelnikow** [comment]
Ooh, possibly useful: https://stackoverflow.com/a/49907438/893113

**Comment by paulmelnikow** [comment]
Fabulous! Thanks for reviewing. I know this was kind of a bear. 🐻

Merge conflict sorted.

**Pull request merged**